File: app/src/main/java/com/mancode/financetracker/ui/balances/AddBalanceFragment.java
package com.mancode.financetracker.ui.balances;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.ViewModelProviders;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkManager;

import com.google.android.material.textfield.TextInputLayout;
import com.mancode.financetracker.AddItemFragment;
import com.mancode.financetracker.R;
import com.mancode.financetracker.database.entity.BalanceEntity;
import com.mancode.financetracker.database.viewmodel.AccountViewModel;
import com.mancode.financetracker.database.viewmodel.BalanceViewModel;
import com.mancode.financetracker.database.views.AccountExtended;
import com.mancode.financetracker.database.workers.UpdateStateWorker;
import com.mancode.financetracker.ui.SetDateView;

import org.threeten.bp.LocalDate;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by Manveru on 07.09.2017.
 */

public class AddBalanceFragment extends AddItemFragment {

    private SetDateView checkDate;
    private AutoCompleteTextView accountDropdown;
    private EditText valueField;

    private Map<String, Integer> accountsMap;
    private ArrayAdapter<String> accountDropdownAdapter;

    // TODO refactor to 1 ViewModel
    private AccountViewModel accountViewModel;
    private BalanceViewModel balanceViewModel;

    public AddBalanceFragment() { }

    static AddBalanceFragment newInstance() {
        return new AddBalanceFragment();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        FragmentActivity activity = getActivity();
        if (activity != null) {
            accountViewModel = ViewModelProviders.of(activity).get(AccountViewModel.class);
            balanceViewModel = ViewModelProviders.of(activity).get(BalanceViewModel.class);
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_add_balance, container, false);

        checkDate = view.findViewById(R.id.tf_balance_check_date);
        accountDropdown = view.findViewById(R.id.dropdownBalanceAccount);
        valueField = view.findViewById(R.id.tf_balance);
        TextInputLayout valueInputLayout = view.findViewById(R.id.balance_value_input_layout);
        valueField.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                if (TextUtils.isEmpty(s)) {
                    valueInputLayout.setError(getString(R.string.error_field_empty));
                } else {
                    valueInputLayout.setError(null);
                }
            }
        });

        List<AccountExtended> accountEntityList = accountViewModel.getAllAccounts().getValue();
        if (accountEntityList != null && getContext() != null) {
            accountsMap = new LinkedHashMap<>();
            for (AccountExtended account : accountEntityList) {
                accountsMap.put(account.accountName, account.id);
            }
            accountDropdownAdapter = new ArrayAdapter<>(
                    getContext(),
                    R.layout.dropdown_menu_popup_item,
                    accountsMap.keySet().toArray(new String[0])
            );
        }
        accountDropdown.setAdapter(accountDropdownAdapter);
        accountDropdown.setText(accountDropdownAdapter.getItem(0), false);

        Toolbar toolbar = view.findViewById(R.id.add_balance_toolbar);
        toolbar.setOnMenuItemClickListener(item -> {
            if (item.getItemId() == R.id.action_menu_save) {
                LocalDate date = checkDate.getDate();
                Integer accountId = accountsMap.get(accountDropdown.getText().toString());
                String valueString = valueField.getText().toString();

                if (!TextUtils.isEmpty(valueString) && accountId != null) {
                    double value = Double.parseDouble(valueString);
                    BalanceEntity balance = new BalanceEntity(
                            0, // not set
                            date,
                            accountId,
                            value,
                            true
                    );
                    balanceViewModel.insert(balance);
                    OneTimeWorkRequest request =
                            new OneTimeWorkRequest.Builder(UpdateStateWorker.class).build();
                    WorkManager.getInstance().enqueue(request);
                    dismiss();
                } else if (TextUtils.isEmpty(valueString)) {
                    valueInputLayout.setError(getString(R.string.error_field_empty));
                } else throw new IllegalStateException("accountId is null!");
            }
            return false;
        });
        toolbar.inflateMenu(R.menu.menu_dialog);
        toolbar.setNavigationOnClickListener(v -> dismiss());
        toolbar.setNavigationIcon(R.drawable.ic_close_white_24dp);

        return view;
    }
}

File: app/src/main/java/com/mancode/financetracker/ui/balances/AddBalanceFragment.kt
package com.mancode.financetracker.ui.balances

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.NavController
import androidx.navigation.fragment.findNavController
import androidx.navigation.fragment.navArgs
import com.mancode.financetracker.R
import com.mancode.financetracker.database.entity.BalanceEntity
import com.mancode.financetracker.database.pojos.BalanceExtended
import com.mancode.financetracker.databinding.FragmentAddBalanceBinding
import com.mancode.financetracker.ui.BalanceInputView
import com.mancode.financetracker.ui.hideKeyboard
import com.mancode.financetracker.utils.InjectorUtils
import com.mancode.financetracker.viewmodel.AddBalancesViewModel
import com.mancode.financetracker.workers.runUpdateWorker
import org.threeten.bp.LocalDate

private val ViewGroup.views: List<View>
    get() = (0 until childCount).map { getChildAt(it) }

class AddBalanceFragment : Fragment(R.layout.fragment_add_balance) {

    private var _binding: FragmentAddBalanceBinding? = null
    private val binding get() = _binding!!

    lateinit var navController: NavController

    private val args: AddBalanceFragmentArgs by navArgs()
    private val viewModel: AddBalancesViewModel by viewModels {
        InjectorUtils.provideAddBalancesViewModelFactory(requireContext(), args.balanceDate
                ?: LocalDate.now())
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        _binding = FragmentAddBalanceBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        navController = findNavController()

        viewModel.accounts.observe(viewLifecycleOwner, { accounts ->
            for (account in accounts) {
                val balanceView = BalanceInputView(requireContext())
                balanceView.setAccount(account)
                binding.container.addView(balanceView)
                if (viewModel.balances.value != null) balanceView.updateWidget(viewModel.balances.value!!)
            }
        })

        viewModel.balances.observe(viewLifecycleOwner, { balances ->
            if (binding.container.childCount > 0) {
                for (balanceView in binding.container.views) {
                    (balanceView as BalanceInputView).updateWidget(balances)
                }
            }
        })

        binding.selectAll.setOnCheckedChangeListener { _, isChecked ->
            for (balanceView in binding.container.views) {
                (balanceView as BalanceInputView).setActive(isChecked)
            }
        }
        if (args.balanceDate == null) binding.selectAll.isChecked = true
        binding.balanceDate.date = viewModel.date
        binding.balanceDate.addDateSetListener {
            val action = AddBalanceFragmentDirections.actionAddBalanceFragmentSelf(binding.balanceDate.date)
            findNavController().navigate(action)
        }

        binding.toolbar.setOnMenuItemClickListener { item ->
            if (item.itemId == R.id.action_menu_save) {

                var noneActive = true
                val accountIds = mutableListOf<Int>()
                for (balanceView in binding.container.views) {
                    val balanceInputView = balanceView as BalanceInputView
                    if (!balanceInputView.isActive() && !balanceInputView.deleted) {
                        continue
                    }
                    noneActive = false
                    val balance = BalanceEntity(
                            balanceInputView.balanceId ?: 0,
                            viewModel.date,
                            balanceInputView.getAccountId(),
                            balanceInputView.getValue(),
                            true
                    )
                    if (balanceInputView.deleted) {
                        viewModel.removeBalance(balance)
                        continue
                    } else {
                        viewModel.addEditBalance(balance)
                    }
                    accountIds.add(balance.accountId)
                }

                if (binding.container.childCount == 0 || noneActive) {
                    Toast.makeText(activity, getString(R.string.error_no_accounts_selected), Toast.LENGTH_SHORT).show()
                } else {
                    requireContext().runUpdateWorker(accountIds.toIntArray(), viewModel.date)
                    dismiss()
                }
            }
            false
        }
        with(binding) {
            toolbar.inflateMenu(R.menu.menu_dialog)
            toolbar.setNavigationOnClickListener { dismiss() }
            toolbar.setNavigationIcon(R.drawable.ic_close_white_24dp)
        }
    }

    override fun onDestroyView() {
        _binding = null
        super.onDestroyView()
    }

    private fun dismiss() {
        hideKeyboard()
        navController.navigateUp()
    }
}

private fun BalanceInputView.updateWidget(balances: List<BalanceExtended>) {
    for (balance in balances) {
        if (getAccountId() == balance.accountId) {
            if (balance.fixed) {
                setValue(balance.value) // TODO needs validation?
                setExisting()
            } else {
                setHint(balance.value)
            }
            balanceId = balance.id
            break
        }
    }
}
